File: app/src/main/java/com/easyfitness/DAO/DAOExerciseInProgram.java
package com.easyfitness.DAO;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.util.Log;
import android.widget.Toast;
import java.util.ArrayList;
import java.util.Date;

public class DAOExerciseInProgram extends DAOBase {

    // Contacts table name
    public static final String TABLE_NAME = "EFExerciseInProgram";
    public static final String KEY = "_id";
    public static final String DATE = "date";
    private static final String TIME = "time";

    public static final String EXERCISE = "machine";
    public static final String PROFIL_KEY = "profil_id";
    public static final String MACHINE_KEY = "machine_id";
    public static final String NOTES = "notes";
    public static final String TYPE = "type";

    // Specific to Strength
    public static final String SERIE = "serie";
    public static final String REPETITION = "repetition";
    public static final String WEIGHT = "poids";
    public static final String UNIT = "unit"; // 0:kg 1:lbs

    // Specific to Cardio
    public static final String DISTANCE = "distance";
    public static final String DURATION = "duration";
    public static final String DISTANCE_UNIT = "distance_unit"; // 0:km 1:mi

    // Specific to STATIC
    public static final String SECONDS = "seconds";
    //rest between exercises
    private static final String REST_SECONDS = "rest_seconds";
    private static final String PROGRAM_ID = "program_id";
    private static final String ORDER_EXECUTION = "order_in_program";

    public static final String TABLE_CREATE = "CREATE TABLE " + TABLE_NAME
        + " (" + KEY + " INTEGER PRIMARY KEY AUTOINCREMENT, " + EXERCISE + " TEXT, "
        + REST_SECONDS + " INTEGER, " + SERIE + " INTEGER, "
        + REPETITION + " INTEGER, " + WEIGHT + " REAL, " + PROFIL_KEY
        + " INTEGER, " + UNIT + " INTEGER, " + NOTES + " TEXT, " + MACHINE_KEY
        + " INTEGER," + TIME + " TEXT," + DISTANCE + " REAL, " + DURATION + " TEXT, "
        + TYPE + " INTEGER, " + SECONDS + " INTEGER, " + DISTANCE_UNIT + " INTEGER, "
        + PROGRAM_ID + " INTEGER, " + ORDER_EXECUTION + " INTEGER);";

//    private String allFieldsExceptProgramId = REST_SECONDS + "," + EXERCISE + "," + SERIE + ","
//        + REPETITION + "," + WEIGHT + "," + WEIGHT + "," + PROFIL_KEY + ","
//        + UNIT + "," + NOTES + "," + MACHINE_KEY + "," + TIME + "," + DISTANCE + "," +
//        DURATION + "," + TYPE + "," + SECONDS + "," + DISTANCE_UNIT+ ","+ORDER_EXECUTION;
//    public static final String TABLE_DROP = "DROP TABLE IF EXISTS "
//        + TABLE_NAME + ";";

    protected Profile mProfile = null;
    protected Cursor mCursor = null;
    protected Context mContext;

    public DAOExerciseInProgram(Context context) {
        super(context);
        mContext = context;
    }

    public void setProfile(Profile pProfile) {
        mProfile = pProfile;
    }

    public Cursor getCursor() {
        return mCursor;
    }

    public int getCount() {
        String countQuery = "SELECT " + KEY + " FROM " + TABLE_NAME;
        open();
        SQLiteDatabase db = this.getReadableDatabase();
        Cursor cursor = db.rawQuery(countQuery, null);
        int value = cursor.getCount();
        cursor.close();
        close();
        return value;
    }

    /**
     * @param pMachine Machine name
     * @return id of the added record, -1 if error
     */
    public long addRecord(long order, long programId, int restSeconds, String pMachine, int pType, int pSerie, int pRepetition, float pPoids, Profile pProfile, int pUnit, String pNote,
                          String pTime, float pDistance, long pDuration, int pSeconds, int distance_unit) {
        ContentValues value = new ContentValues();
        long new_id = -1;
        long machine_key = -1;

        DAOExerciseInProgram daoExerciseInProgram = new DAOExerciseInProgram(mContext);
        if (daoExerciseInProgram.exerciseExists(pMachine)) {
            return new_id;
        }

        value.put(DAOExerciseInProgram.PROGRAM_ID, programId);
        value.put(DAOExerciseInProgram.REST_SECONDS, restSeconds);
        value.put(DAOExerciseInProgram.EXERCISE, pMachine);
        value.put(DAOExerciseInProgram.SERIE, pSerie);
        value.put(DAOExerciseInProgram.REPETITION, pRepetition);
        value.put(DAOExerciseInProgram.WEIGHT, pPoids);
        value.put(DAOExerciseInProgram.PROFIL_KEY, pProfile.getId());
        value.put(DAOExerciseInProgram.UNIT, pUnit);
        value.put(DAOExerciseInProgram.NOTES, pNote);
        value.put(DAOExerciseInProgram.MACHINE_KEY, machine_key);
        value.put(DAOExerciseInProgram.TIME, pTime);
        value.put(DAOExerciseInProgram.DISTANCE, pDistance);
        value.put(DAOExerciseInProgram.DURATION, pDuration);
        value.put(DAOExerciseInProgram.TYPE, pType);
        value.put(DAOExerciseInProgram.SECONDS, pSeconds);
        value.put(DAOExerciseInProgram.DISTANCE_UNIT, distance_unit);
        value.put(DAOExerciseInProgram.ORDER_EXECUTION, order);
        SQLiteDatabase db = open();
        new_id = db.insert(DAOExerciseInProgram.TABLE_NAME, null, value);
        close();
        return new_id;
    }

    private boolean exerciseExists(String name) {
        ExerciseInProgram lMach = getRecord(name);
        return lMach != null;
    }

    private ExerciseInProgram getRecord(String pName) {
        SQLiteDatabase db = this.getReadableDatabase();
        mCursor = null;
        mCursor = db.query(TABLE_NAME, new String[]{EXERCISE}, EXERCISE + "=?",
            new String[]{pName}, null, null, null, null);
        if (mCursor != null)
            mCursor.moveToFirst();

        if (mCursor.getCount() == 0)
            return null;
        DAOProfil lDAOProfil = new DAOProfil(mContext);
        Profile lProfile = lDAOProfil.getProfil(mCursor.getLong(mCursor.getColumnIndex(DAOFonte.PROFIL_KEY)));
        ExerciseInProgram value = new ExerciseInProgram(
            mCursor.getInt(mCursor.getColumnIndex(REST_SECONDS)),
            mCursor.getString(mCursor.getColumnIndex(EXERCISE)),
            mCursor.getInt(mCursor.getColumnIndex(SERIE)),
            mCursor.getInt(mCursor.getColumnIndex(REPETITION)),
            mCursor.getFloat(mCursor.getColumnIndex(WEIGHT)),
            lProfile,
            mCursor.getInt(mCursor.getColumnIndex(UNIT)),
            mCursor.getString(mCursor.getColumnIndex(NOTES)),
            mCursor.getInt(mCursor.getColumnIndex(MACHINE_KEY)),
            mCursor.getString(mCursor.getColumnIndex(TIME))
        );

        value.setId(mCursor.getLong(0));
        // return value
        mCursor.close();
        close();
        return value;
    }

    public void deleteRecord(long id) {
        SQLiteDatabase db = this.getWritableDatabase();
        db.delete(TABLE_NAME, KEY + " = ?", new String[]{String.valueOf(id)});
        db.close();
    }

    public IRecord getRecord(long id) {
        String selectQuery = "SELECT  * FROM " + TABLE_NAME + " WHERE " + KEY + "=" + id;

        mCursor = getRecordsListCursor(selectQuery);
        if (mCursor.moveToFirst()) {
            //Get Profile
            DAOProfil lDAOProfil = new DAOProfil(mContext);
            Profile lProfile = lDAOProfil.getProfil(mCursor.getLong(mCursor.getColumnIndex(DAOExerciseInProgram.PROFIL_KEY)));

            long machine_key = -1;

            //Test is Machine exists. If not create it.
            DAOMachine lDAOMachine = new DAOMachine(mContext);
            if (mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.MACHINE_KEY)) == null) {
                machine_key = lDAOMachine.addMachine(mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.EXERCISE)), "", DAOMachine.TYPE_FONTE, "", false, "");
            } else {
                machine_key = mCursor.getLong(mCursor.getColumnIndex(DAOExerciseInProgram.MACHINE_KEY));
            }

            IRecord value = null;

            if (mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.TYPE)) == DAOMachine.TYPE_FONTE) {
                value = new ExerciseInProgram(mCursor.getColumnIndex(DAOExerciseInProgram.SECONDS),
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.EXERCISE)),
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.SERIE)),
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.REPETITION)),
                    mCursor.getFloat(mCursor.getColumnIndex(DAOExerciseInProgram.WEIGHT)),
                    lProfile,
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.UNIT)),
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.NOTES)),
                    machine_key,
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.TIME)));
            } else if (mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.TYPE)) == DAOMachine.TYPE_STATIC) {
                value = new StaticExercise(new Date(),
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.EXERCISE)),
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.SERIE)),
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.SECONDS)),
                    mCursor.getFloat(mCursor.getColumnIndex(DAOExerciseInProgram.WEIGHT)),
                    lProfile,
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.UNIT)),
                    machine_key,
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.TIME)));
            } else {
                value = new Cardio(new Date(),
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.EXERCISE)),
                    mCursor.getFloat(mCursor.getColumnIndex(DAOExerciseInProgram.DISTANCE)),
                    mCursor.getLong(mCursor.getColumnIndex(DAOExerciseInProgram.DURATION)),
                    lProfile,
                    mCursor.getString(mCursor.getColumnIndex(DAOExerciseInProgram.TIME)),
                    mCursor.getInt(mCursor.getColumnIndex(DAOExerciseInProgram.DISTANCE_UNIT)));
            }

            value.setId(mCursor.getLong(mCursor.getColumnIndex(DAOExerciseInProgram.KEY)));
            return value;
        } else {
            return null;
        }
    }

    // Getting All Records
    private Cursor getRecordsListCursor(String pRequest) {
        SQLiteDatabase db = this.getReadableDatabase();
        return db.rawQuery(pRequest, null);
    }

    // Getting All Machines
    public String[] getAllMachines(Profile pProfile) {
        SQLiteDatabase db = this.getReadableDatabase();
        mCursor = null;

        // Select All Machines
        String selectQuery = "SELECT DISTINCT " + EXERCISE + " FROM "
            + TABLE_NAME + "  WHERE " + PROFIL_KEY + "=" + pProfile.getId() + " ORDER BY " + KEY + " ASC";
        mCursor = db.rawQuery(selectQuery, null);

        int size = mCursor.getCount();

        String[] valueList = new String[size];

        // looping through all rows and adding to list
        if (mCursor.moveToFirst()) {
            int i = 0;
            do {
                String value = mCursor.getString(0);
                valueList[i] = value;
                i++;
            } while (mCursor.moveToNext());
        }
        close();
        // return value list
        return valueList;
    }

    public Cursor getTop3DatesRecords(Profile pProfile) {
        if (pProfile == null)
            return null;
        String selectQuery = "SELECT * FROM " + TABLE_NAME
            + " WHERE " + PROFIL_KEY + "=" + pProfile.getId()
            + " AND " + KEY + " IN (SELECT DISTINCT " + KEY + " FROM " + TABLE_NAME + " WHERE " + PROFIL_KEY + "=" + pProfile.getId() + " ORDER BY " + KEY + " ASC LIMIT 10)"
            + " ORDER BY " + KEY + " ASC";
        return getRecordsListCursor(selectQuery);
    }

    /**
     * @return the last record for a profile p
     */
    public IRecord getLastRecord(Profile pProfile) {
        SQLiteDatabase db = this.getReadableDatabase();
        mCursor = null;
        IRecord lReturn = null;
        String selectQuery = "SELECT MAX(" + KEY + ") FROM " + TABLE_NAME
            + " WHERE " + PROFIL_KEY + "=" + pProfile.getId();
        mCursor = db.rawQuery(selectQuery, null);
        if (mCursor.moveToFirst()) {
            try {
                long value = mCursor.getLong(0);
                lReturn = getRecord(value);
            } catch (NumberFormatException e) {
                lReturn = null; // Return une valeur
            }
        }
        close();
        return lReturn;
    }

    public void closeCursor() {
        if (mCursor != null) mCursor.close();
    }

    public ArrayList<ARecord> getAllExerciseInProgramArray() {
        String selectQuery = "SELECT * FROM " + TABLE_NAME + ";";//+ " ORDER BY "
        // + KEY + " ASC;";
        return getExerciseList(selectQuery);
    }

    public ArrayList<ARecord> getAllExerciseInProgram(Long programId) {
        String selectQuery = "SELECT * FROM " + TABLE_NAME + " WHERE " + PROGRAM_ID + " = " + programId
            + " ORDER BY " + KEY + " DESC;";
        return getExerciseList(selectQuery);
    }

    private ArrayList<ARecord> getExerciseList(String pRequest) {
        ArrayList<ARecord> valueList = new ArrayList<>();
        SQLiteDatabase db = this.getReadableDatabase();
        mCursor = null;
        try {
            mCursor = db.rawQuery(pRequest, null);
        } catch (Exception ex) {
            Log.e("Err MU", "EX " + ex);
            Toast.makeText(mContext, "Ex" + ex, Toast.LENGTH_LONG).show();
        }
        // looping through all rows and adding to list
        if (mCursor.moveToFirst()) {
            do {
                ExerciseInProgram value = new ExerciseInProgram(
                    mCursor.getInt(mCursor.getColumnIndex(REST_SECONDS)),
                    mCursor.getString(mCursor.getColumnIndex(EXERCISE)),
                    mCursor.getInt(mCursor.getColumnIndex(SERIE)),
                    mCursor.getInt(mCursor.getColumnIndex(REPETITION)),
                    mCursor.getInt(mCursor.getColumnIndex(WEIGHT)),
                    mProfile,
                    mCursor.getInt(mCursor.getColumnIndex(UNIT)),
                    mCursor.getString(mCursor.getColumnIndex(NOTES)),
                    mCursor.getInt(mCursor.getColumnIndex(MACHINE_KEY)),
                    mCursor.getString(mCursor.getColumnIndex(TIME))
                );

                value.setId(mCursor.getLong(mCursor.getColumnIndex(KEY)));
                valueList.add(value);
            } while (mCursor.moveToNext());
        }
        return valueList;
    }
}

File: app/src/main/java/com/easyfitness/DAO/DAOExerciseInProgram.kt
package com.easyfitness.DAO

import android.content.ContentValues
import android.content.Context
import android.database.Cursor
import android.widget.Toast
import timber.log.Timber
import java.util.*

class DAOExerciseInProgram(var mContext: Context) : DAOBase(mContext) {
    //    private String allFieldsExceptProgramId = REST_SECONDS + "," + EXERCISE + "," + SERIE + ","
    //        + REPETITION + "," + WEIGHT + "," + WEIGHT + "," + PROFIL_KEY + ","
    //        + UNIT + "," + NOTES + "," + MACHINE_KEY + "," + TIME + "," + DISTANCE + "," +
    //        DURATION + "," + TYPE + "," + SECONDS + "," + DISTANCE_UNIT+ ","+ORDER_EXECUTION;
    var mProfile: Profile? = null
    var cursor: Cursor? = null

    fun setProfile(pProfile: Profile?) {
        mProfile = pProfile
    }

    val count: Int
        get() {
            val countQuery = "SELECT $KEY FROM $TABLE_NAME"
            open()
            val db = this.readableDatabase
            val cursor = db.rawQuery(countQuery, null)
            val value = cursor.count
            cursor.close()
            close()
            return value
        }

    /**
     * @param pMachine Machine name
     * @return id of the added record, -1 if error
     */
    fun addRecord(order: Long, programId: Long, restSeconds: Int, pMachine: String, pType: Int, pSerie: Int, pRepetition: Int, pPoids: Float, pProfile: Profile, pUnit: Int, pNote: String?,
                  pTime: String?, pDistance: Float, pDuration: Long, pSeconds: Int, distance_unit: Int): Long {
        val value = ContentValues()
        var newId: Long = -1
        val machineKey: Long = -1
        val daoExerciseInProgram = DAOExerciseInProgram(mContext)
        if (daoExerciseInProgram.exerciseExists(pMachine)) {
            return newId
        }
        value.put(PROGRAM_ID, programId)
        value.put(REST_SECONDS, restSeconds)
        value.put(EXERCISE, pMachine)
        value.put(SERIE, pSerie)
        value.put(REPETITION, pRepetition)
        value.put(WEIGHT, pPoids)
        value.put(PROFIL_KEY, pProfile.id)
        value.put(UNIT, pUnit)
        value.put(NOTES, pNote)
        value.put(MACHINE_KEY, machineKey)
        value.put(TIME, pTime)
        value.put(DISTANCE, pDistance)
        value.put(DURATION, pDuration)
        value.put(TYPE, pType)
        value.put(SECONDS, pSeconds)
        value.put(DISTANCE_UNIT, distance_unit)
        value.put(ORDER_EXECUTION, order)
        val db = open()
        newId = db.insert(TABLE_NAME, null, value)
        close()
        return newId
    }

    private fun exerciseExists(name: String): Boolean {
        val lMach = getRecord(name)
        return lMach != null
    }

    private fun getRecord(pName: String): ExerciseInProgram? {
        val db = this.readableDatabase
        cursor = null
        cursor = db.query(TABLE_NAME, arrayOf(EXERCISE), "$EXERCISE=?", arrayOf(pName), null, null, null, null)
        if (cursor != null) cursor!!.moveToFirst()
        if (0 == cursor!!.count) { close()
            cursor!!.close()
            return null}
        val lDAOProfil = DAOProfil(mContext)
        val lProfile = lDAOProfil.getProfil(cursor!!.getLong(cursor!!.getColumnIndex(DAOFonte.PROFIL_KEY)))
        val value = ExerciseInProgram(
            cursor!!.getInt(cursor!!.getColumnIndex(REST_SECONDS)),
            cursor!!.getString(cursor!!.getColumnIndex(EXERCISE)),
            cursor!!.getInt(cursor!!.getColumnIndex(SERIE)),
            cursor!!.getInt(cursor!!.getColumnIndex(REPETITION)),
            cursor!!.getFloat(cursor!!.getColumnIndex(WEIGHT)),
            lProfile,
            cursor!!.getInt(cursor!!.getColumnIndex(UNIT)),
            cursor!!.getString(cursor!!.getColumnIndex(NOTES)),
            cursor!!.getInt(cursor!!.getColumnIndex(MACHINE_KEY)).toLong(),
            cursor!!.getString(cursor!!.getColumnIndex(TIME)),
            cursor!!.getInt(cursor!!.getColumnIndex(TYPE))
        )
        value.setId(cursor!!.getLong(0))

        close()
        cursor!!.close()
        return value
    }

    fun deleteRecord(id: Long) {
        val db = this.writableDatabase
        db.delete(TABLE_NAME, "$KEY = ?", arrayOf(id.toString()))
        db.close()
    }

    // Getting All Records
    private fun getRecordsListCursor(pRequest: String): Cursor {
        val db = this.readableDatabase
        return db.rawQuery(pRequest, null)
    }

    fun getAllExerciseInProgramToRecord(programId: Long): Cursor? {
        val selectQuery = ("SELECT * FROM " + TABLE_NAME
            + " WHERE " + PROGRAM_ID + "=" + programId
            + " AND " + KEY + " IN (SELECT DISTINCT " + KEY + " FROM " + TABLE_NAME + " WHERE " + PROGRAM_ID + "=" + programId + " ORDER BY " + ORDER_EXECUTION + " ASC)"
            + " ORDER BY " + KEY + " ASC")
        return getRecordsListCursor(selectQuery)
    }

    fun getAllExerciseInProgramAsList(programId: Long): ArrayList<ARecord> {
        val selectQuery = ("SELECT * FROM " + TABLE_NAME + " WHERE " + PROGRAM_ID + " = " + programId
            + " ORDER BY " + ORDER_EXECUTION + " ASC;")
        return getExerciseListToList(selectQuery)
    }

    fun getAllExerciseInProgram(programId: Long): ArrayList<ExerciseInProgram> {
        val selectQuery = ("SELECT * FROM " + TABLE_NAME + " WHERE " + PROGRAM_ID + " = " + programId
            + " ORDER BY " + ORDER_EXECUTION + " ASC;")
        return getExerciseList(selectQuery)
    }

    private fun getExerciseList(pRequest: String): ArrayList<ExerciseInProgram> {
        val valueList = ArrayList<ExerciseInProgram>()
        val db = this.readableDatabase
        cursor = null
        try {
            cursor = db.rawQuery(pRequest, null)
        } catch (ex: Exception) {
            Timber.e("Err in getExList $ex")
            Toast.makeText(mContext, "Ex$ex", Toast.LENGTH_LONG).show()
        }
        // looping through all rows and adding to list
        if (cursor!!.moveToFirst()) {
            do {
                val value = ExerciseInProgram( //int secRest, String pMachine, int pSerie, int pRepetition, float pPoids,
                    //                             Profile pProfile, int pUnit, String pNote, long pMachineKey, String pTime,
                    //                             int type, int distance, String duration, int seconds, int distanceUnit,
                    //                             long order
                    cursor!!.getInt(cursor!!.getColumnIndex(REST_SECONDS)),
                    cursor!!.getString(cursor!!.getColumnIndex(EXERCISE)),
                    cursor!!.getInt(cursor!!.getColumnIndex(SERIE)),
                    cursor!!.getInt(cursor!!.getColumnIndex(REPETITION)),
                    cursor!!.getInt(cursor!!.getColumnIndex(WEIGHT)).toFloat(),
                    mProfile,
                    cursor!!.getInt(cursor!!.getColumnIndex(UNIT)),
                    cursor!!.getString(cursor!!.getColumnIndex(NOTES)),
                    cursor!!.getInt(cursor!!.getColumnIndex(MACHINE_KEY)).toLong(),
                    cursor!!.getString(cursor!!.getColumnIndex(TIME)),
                    cursor!!.getInt(cursor!!.getColumnIndex(TYPE)),
                    cursor!!.getInt(cursor!!.getColumnIndex(DISTANCE)),
                    cursor!!.getLong(cursor!!.getColumnIndex(DURATION)),
                    cursor!!.getInt(cursor!!.getColumnIndex(SECONDS)),
                    cursor!!.getInt(cursor!!.getColumnIndex(DISTANCE_UNIT)),
                    cursor!!.getLong(cursor!!.getColumnIndex(ORDER_EXECUTION))
                )
                value.setId(cursor!!.getLong(cursor!!.getColumnIndex(KEY)))
                valueList.add(value)
            } while (cursor!!.moveToNext())
        }
        cursor!!.close()
        close()
        return valueList
    }

    private fun getExerciseListToList(pRequest: String): ArrayList<ARecord> {
        val valueList = ArrayList<ARecord>()
        val db = this.readableDatabase
        cursor = null
        try {
            cursor = db.rawQuery(pRequest, null)
        } catch (ex: Exception) {
            Timber.e("Err in getExToList $ex")
            Toast.makeText(mContext, "Ex$ex", Toast.LENGTH_LONG).show()
        }
        // looping through all rows and adding to list
        if (cursor!!.moveToFirst()) {
            do {
                val value = ExerciseInProgram(
                    cursor!!.getInt(cursor!!.getColumnIndex(REST_SECONDS)),
                    cursor!!.getString(cursor!!.getColumnIndex(EXERCISE)),
                    cursor!!.getInt(cursor!!.getColumnIndex(SERIE)),
                    cursor!!.getInt(cursor!!.getColumnIndex(REPETITION)),
                    cursor!!.getInt(cursor!!.getColumnIndex(WEIGHT)).toFloat(),
                    mProfile,
                    cursor!!.getInt(cursor!!.getColumnIndex(UNIT)),
                    cursor!!.getString(cursor!!.getColumnIndex(NOTES)),
                    cursor!!.getInt(cursor!!.getColumnIndex(MACHINE_KEY)).toLong(),
                    cursor!!.getString(cursor!!.getColumnIndex(TIME)),
                    cursor!!.getInt(cursor!!.getColumnIndex(TYPE))
                    )
                value.setId(cursor!!.getLong(cursor!!.getColumnIndex(KEY)))
                valueList.add(value)
            } while (cursor!!.moveToNext())
        }
        close()
        cursor!!.close()
        return valueList
    }

    fun updateString(exerciseInProgram: ExerciseInProgram , field: String, newValue: String):Int {
            val db = this.writableDatabase
            val value = ContentValues()
            value.put(field, newValue)
            return db.update(TABLE_NAME, value, "$KEY = ?", arrayOf(exerciseInProgram.id.toString()))
    }

    companion object {
        // Contacts table name
        const val TABLE_NAME = "EFExerciseInProgram"
        const val KEY = "_id"
        const val DATE = "date"
        private const val TIME = "time"
        const val EXERCISE = "machine"
        const val PROFIL_KEY = "profil_id"
        private const val MACHINE_KEY = "machine_id"
        const val NOTES = "notes"
        const val TYPE = "type"

        // Specific to Strength
        const val SERIE = "serie"
        const val REPETITION = "repetition"
        const val WEIGHT = "poids"
        const val UNIT = "unit" // 0:kg 1:lbs

        // Specific to Cardio
        const val DISTANCE = "distance"
        const val DURATION = "duration"
        const val DISTANCE_UNIT = "distance_unit" // 0:km 1:mi

        // Specific to STATIC
        const val SECONDS = "seconds"

        //rest between exercises
        private const val REST_SECONDS = "rest_seconds"
        private const val PROGRAM_ID = "program_id"
        const val ORDER_EXECUTION = "order_in_program"
        const val TABLE_CREATE = ("CREATE TABLE " + TABLE_NAME
            + " (" + KEY + " INTEGER PRIMARY KEY AUTOINCREMENT, " + EXERCISE + " TEXT, "
            + REST_SECONDS + " INTEGER, " + SERIE + " INTEGER, "
            + REPETITION + " INTEGER, " + WEIGHT + " REAL, " + PROFIL_KEY
            + " INTEGER, " + UNIT + " INTEGER, " + NOTES + " TEXT, " + MACHINE_KEY
            + " INTEGER," + TIME + " TEXT," + DISTANCE + " REAL, " + DURATION + " TEXT, "
            + TYPE + " INTEGER, " + SECONDS + " INTEGER, " + DISTANCE_UNIT + " INTEGER, "
            + PROGRAM_ID + " INTEGER, " + ORDER_EXECUTION + " INTEGER);")
    }

}
